kit/utils/observer.go

```diff
@@ -0,0 +1,62 @@
+package utils
+
+import (
+	"gopkg.in/ibrt/go-xerror.v2/xerror"
+	"sync"
+)
+
+const (
+	// ErrorSubscriber is return when an error occurs while processing an event in a subscriber.
+	ErrorSubscriber = "subscriber error"
+)
+
+// Observer describes the observer pattern.
+type Observer interface {
+	Publish(eventName string, eventData interface{}) error
+	Subscribe(eventName string, eventSubscriber EventSubscriber)
+}
+
+// EventSubscriber is a callback function for observed events.
+type EventSubscriber func(eventName string, eventData interface{}) error
```

Comment:
It's a more flexible API, now you can have the same subscriber function for different event types.